griptape/drivers/vector/pgvector_vector_store_driver.py

```diff
@@ -0,0 +1,170 @@
+            self.engine = create_engine(self.connection_string, **self.create_engine_params)
+
+        self._Model = vector_model_factory(table_name)
+
+        if install_uuid_extension:
+            self.engine.execute('CREATE EXTENSION IF NOT EXISTS "uuid-ossp";')
+
+        if install_vector_extension:
+            self.engine.execute('CREATE EXTENSION IF NOT EXISTS "vector";')
+
+        if create_schema:
+            self._Model.metadata.create_all(self.engine)
+
+    def upsert_vector(
+        self,
+        vector: list[float],
+        vector_id: Optional[str] = None,
+        namespace: Optional[str] = None,
+        meta: Optional[dict] = None,
+        **kwargs
+    ) -> str:
+        """Inserts or updates a vector in the collection."""
+        with Session(self.engine) as session:
+            obj = self._Model(
+                id=vector_id,
+                vector=vector,
+                namespace=namespace,
+                meta=meta,
+            )
+
+            obj = session.merge(obj)
+            session.commit()
+
+            return str(obj.id)
+
+    def load_entry(self, vector_id: str, namespace: Optional[str] = None) -> BaseVectorStoreDriver.Entry:
+        """Retrieves a specific vector entry from the collection based on its identifier and optional namespace."""
+        with Session(self.engine) as session:
+            result = session.get(self._Model, vector_id)
+
+            return BaseVectorStoreDriver.Entry(
+                id=result.id,
+                vector=result.vector,
+                namespace=result.namespace,
+                meta=result.meta,
+            )
+
+    def load_entries(self, namespace: Optional[str] = None) -> list[BaseVectorStoreDriver.Entry]:
+        """Retrieves all vector entries from the collection, optionally filtering to only
+        those that match the provided namespace.
+        """
+        with Session(self.engine) as session:
+            query = session.query(self._Model)
+            if namespace:
+                query = query.filter_by(namespace=namespace)
+
+            results = query.all()
+
+            return [
+                BaseVectorStoreDriver.Entry(
+                    id=str(result.id),
+                    vector=result.vector,
+                    namespace=result.namespace,
+                    meta=result.meta,
+                )
+                for result in results
+            ]
+
+    def query(
+        self,
+        query: str,
+        count: Optional[int] = BaseVectorStoreDriver.DEFAULT_QUERY_COUNT,
+        namespace: Optional[str] = None,
+        include_vectors: bool = False,
+        distance_metric: str = "cosine_distance",
+        **kwargs
+    ) -> list[BaseVectorStoreDriver.QueryResult]:
+        """Performs a search on the collection to find vectors similar to the provided input vector,
+        optionally filtering to only those that match the provided namespace.
+        """
+        distance_metrics = {
```

Comment:
Any way to move those to a constant at the top of the class? Should be possible if those be accessed as class properties of `VectorModel` directly.

Reply:
Not sure if this is doable without lots of magic as the dict values depend on the dynamic `_model` being built. The comparison methods themselves are returned by a factory inside `sqlalchemy.Vector` and must be instantiated for this to work according to my experimentation.